src/main/java/org/folio/edge/sip2/repositories/CirculationRepository.java

```diff
@@ -107,54 +115,95 @@
-          if (resource.getResource() != null) {
-            final String dueDateString = resource.getResource().getString("dueDate", null);
-            if (dueDateString == null) {
-              dueDate = null;
-            } else {
-              // Need to convert to the tenant local timezone
-              dueDate = OffsetDateTime.from(Utils.getFolioDateTimeFormatter().parse(dueDateString));
-            }
+    return loginFuture
+        .compose(login -> {
+          final JsonObject body = new JsonObject()
+              .put("itemBarcode", itemIdentifier)
+              .put("userBarcode", lmsPatronIdentifier[0])
+              .put("servicePointId", sessionData.getScLocation());
+
+          final Map<String, String> headers = getBaseHeaders();
+
+          final CheckoutRequestData checkoutRequestData =
+              new CheckoutRequestData(body, headers, sessionData);
+
+          // This means we tried to log in as the patron and failed in some way
+          if (login != null && !login.getErrorMessages().isEmpty()) {
+            return Future.succeededFuture(CheckoutResponse.builder()
+                .ok(FALSE)
+                .renewalOk(FALSE)
+                .magneticMedia(null)
+                .desensitize(FALSE)
+                .transactionDate(OffsetDateTime.now(clock))
+                .institutionId(institutionId)
+                .patronIdentifier(patronIdentifier)
+                .itemIdentifier(itemIdentifier)
+                .titleIdentifier(UNKNOWN)
+                .dueDate(OffsetDateTime.now(clock))
+                .screenMessage(login.getErrorMessages())
+                .build());
           } else {
-            dueDate = null;
+            final Future<IResource> result = resourceProvider.createResource(checkoutRequestData);
+
+            return result
+                .otherwise(Utils::handleErrors)
+                .map(resource -> {
+                  final Optional<JsonObject> response = Optional.ofNullable(resource.getResource());
+                  // The problem here is that the checkout has already succeeded, but something
+                  // could be wrong with the returned data. The odds of this are low. The "dueDate"
+                  // field is not required, so a loan could be missing one. I am not sure
+                  // what that means for SIP2 where the due date field is required. Setting it to
+                  // null is probably wrong and will likely break the template when building the
+                  // SIP2 response.
+                  final OffsetDateTime dueDate = response
+                      .map(v -> v.getString("dueDate"))
+                      .map(v -> OffsetDateTime.from(Utils.getFolioDateTimeFormatter().parse(v)))
+                      .orElse(null);
+
+                  return CheckoutResponse.builder()
+                      .ok(Boolean.valueOf(response.isPresent()))
+                      .renewalOk(FALSE)
```

Comment:
RenewalOk = false.. False is because our SIP2 does not support renewal yet or because the user cannot renew? What should the value here be? If the former then one day we implement renewal then we will need to remember to go back to this line of code to flip it back to true? I think it should be based on the ability of the user to renew, and not based on SIP2 implementation, but I could be wrong.

Reply:
If `renewalOk` is set to `TRUE`, it is supposed to mean that the item was already checked out by the patron and the ACS did a renewal instead. `FALSE` means that it was a regular check out.

I don't think FOLIO checkout does a renew if the item is already checked out to the patron, so this is something we'd have to implement in this module (or propose for circulation). For now, since we can't do this, I just set it to `FALSE`.

I'm not sure how or if this is related to ACS/SC renewal policy. Since there are separate renew/renew all SIP commands that can be enabled/disabled via supported messages and the doc says the renewal policy fields used to be called "renewal allowed/ok" in SIP1, I would think there is some relation here.

My theory is that on checkout, if SC renewal policy is set to Y and ACS renewal policy is set to Y, we should try to renew if the item is checked out to the patron and set `renewalOk` to `TRUE` and this would be the only way it is set to `TRUE`.

Reply:
Yeah, I forgot to consult the SIP2 protocol to understand the renewalOk field logic. Makes sense.